
<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [181]:
import pandas as pd
import psycopg2

In [182]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [183]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [184]:
# текст запроса
query_3_1 = f'''select 
                count(id)
                from public.vacancies
'''

In [185]:
# результат запроса
count_vacancies = pd.read_sql_query(query_3_1, connection)
print(count_vacancies)

   count
0  49197


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\3566143398.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacancies = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [186]:
# текст запроса
query_3_2 = f'''select 
                count(id) 
                from public.employers
'''

In [187]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
print(count_employers)

   count
0  23501


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1831000636.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [188]:
# текст запроса
query_3_3 = f'''select 
                count(id) 
                from public.areas
'''

In [189]:
# результат запроса
count_areas = pd.read_sql_query(query_3_3, connection)
print(count_areas)

   count
0   1362


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1480312434.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_areas = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [190]:
# текст запроса
query_3_4 = f'''select 
                count(id) 
                from public.industries
'''

In [191]:
# результат запроса
count_industries = pd.read_sql_query(query_3_4, connection)
print(count_industries)

   count
0    294


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1882296354.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_industries = pd.read_sql_query(query_3_4, connection)


***

##### Выводы
В таблицах представлены данные о более 49 тыс. вакансий от более 23 тыс. работодателей из 1362 городов(регионов). Количество сфер деятельности работодателей 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [192]:
# текст запроса
query_4_1 = f'''select 
                count(v.id) as cnt,
                a.name as area
                from public.vacancies v
                join public.areas a on a.id = v.area_id
                group by 2
                order by 1 DESC
'''

In [193]:
# результат запроса
vacancies_by_area = pd.read_sql_query(query_4_1, connection)
display(vacancies_by_area)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2045252745.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_by_area = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [194]:
# текст запроса
query_4_2 = f'''select 
                count(v.id)
                from public.vacancies v
                where (v.salary_from is not null) or (v.salary_to is not null)
'''

In [195]:
# результат запроса
print(pd.read_sql_query(query_4_2, connection))

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2467670591.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  print(pd.read_sql_query(query_4_2, connection))


   count
0  24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [196]:
# текст запроса
query_4_3 = f'''select 
                round((avg(v.salary_from)), 0) as low,
                round((avg(v.salary_to)), 0) as high
                from public.vacancies v
'''

In [197]:
# результат запроса
salary_values = pd.read_sql_query(query_4_3, connection)
print(salary_values)


       low      high
0  71065.0  110537.0


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\640903058.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_values = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [198]:
# текст запроса
query_4_4 = f'''select 
                v.schedule,
                vv.employment,
                count (v.id)
                from public.vacancies v
                join public.vacancies vv on v.id = vv.id
                group by 1,2
                order by 3 desc
'''

In [199]:
# результат запроса
variants = pd.read_sql_query(query_4_4, connection)
print(variants)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2043299875.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  variants = pd.read_sql_query(query_4_4, connection)


            schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [200]:
# текст запроса
query_4_5 = f'''select 
                count(v.experience),
                v.experience
                from public.vacancies v
                group by v.experience
                order by count(v.experience) asc
'''

In [201]:
# результат запроса
count_by_experience = pd.read_sql_query(query_4_5, connection)
print(count_by_experience)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\792772146.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_by_experience = pd.read_sql_query(query_4_5, connection)


   count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет


***

##### Выводы
1. Основными работодателями, являются компании расположенные в столицах и региональных центрах. Это обусловленно тем, что там находятся штабы крупнейших организаций.
2. Верхний или нижний порог зарплат указан примерно у половины вакансий. Это говорит, о том что зарплата кандидатов по вакансиям без зарплатной вилки, вероятно, обсуждается по результатам собеседований, в сответствии с опытом и навыками претедентов.
3. Уровни средних значений зарплатной вилки от 71 до 110 тыс. рублей. 
4. Большая часть вакансий это работа с классическим 8-ми часовым днем на месте работодателя. Но на втором месте находится удаленная работа с полной занятостью. Связано это с тем, что удаленный формат работы стал актуальным в период пандемии и продлжает набирать популярность, как у соискателей так и у работодателей.
5. Больше половины вакансий требуют опыта от 1 до 3 лет, еще примерно 30% от 3 до 6 лет. Подавляющая часть вакансий для соискателей с опытом.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [202]:
# текст запроса
query_5_1 = f'''select 
                count(v.id),
                e.name
                from public.vacancies v
                join public.employers e on e.id = v.employer_id
                group by 2
                order by count(v.id) desc
                limit 5
'''

In [203]:
# результат запроса
employers = pd.read_sql_query(query_5_1, connection)
print(employers)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2998975191.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers = pd.read_sql_query(query_5_1, connection)


   count           name
0   1933         Яндекс
1    491     Ростелеком
2    444       Тинькофф
3    428           СБЕР
4    331  Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [204]:
# текст запроса
query_5_2= f'''select 
                count(v.id) as count_of_vacancies,
                count(e.id) as count_of_employers,
                a.name
                from areas a
                left join vacancies v on v.area_id = a.id
                left join employers e on e.area = a.id
                where v.id is null
                group by 3
                order by 1 asc, 2 desc
                limit 5
'''

In [205]:
# результат запроса
areas_with_null_vacancies = pd.read_sql_query(query_5_2, connection)
print(areas_with_null_vacancies)


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\245919449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_with_null_vacancies = pd.read_sql_query(query_5_2, connection)


   count_of_vacancies  count_of_employers                name
0                   0                 410              Россия
1                   0                 207           Казахстан
2                   0                  75  Московская область
3                   0                  19  Краснодарский край
4                   0                  18            Беларусь


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [206]:
# текст запроса
query_5_3 = f'''select 
                e.name employer,
                count (distinct v.area_id) 
                from employers e
                join vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
'''

In [207]:
# результат запроса
regions_by_employers = pd.read_sql_query(query_5_3, connection)
print(regions_by_employers)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\825446308.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  regions_by_employers = pd.read_sql_query(query_5_3, connection)


                     employer  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [208]:
# текст запроса
query_5_4 = f'''select 
                count(e.id)
                from employers e
                left join employers_industries ei on ei.employer_id = e.id
                where ei.industry_id is null
'''

In [209]:
# результат запроса
count_employers_with_null_industries = pd.read_sql_query(query_5_4, connection)
print(count_employers_with_null_industries)

   count
0   8419


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2879251951.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers_with_null_industries = pd.read_sql_query(query_5_4, connection)


In [210]:
# дополнительный запрос для вывода части наименований таких работодателей, с целью понять, что это за организации
query_5_4_1 = f'''select 
                name(e.name)
                from employers e
                left join employers_industries ei on ei.employer_id = e.id
                where ei.industry_id is null
                limit 20
'''

In [211]:
employers_with_null_industries = pd.read_sql_query(query_5_4_1, connection)
print(employers_with_null_industries)

                                   name
0                             Вертикаль
1              Нозиков Денис Михайлович
2     МВД России по Республике Башкорто
3     ГКУ «Центр цифровой трансформации
4   Театральный музей им. А.А. Бахрушин
5                              Полтавка
6                           БИТС-Сервис
7                          SiteAnalyzer
8                            Альта-Софт
9                              НПП АММА
10                              МОДИМИО
11                             Found.ee
12                        Рыжиков Д. В.
13                                 НЧТЗ
14    ГКУ Центр занятости населения Пер
15                    СТМ-ЭНЕРГОРЕШЕНИЯ
16                          Энергострой
17                           АИСТ ГЕЙМС
18                            Стройсбыт
19                                Ранкс


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1154506309.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_with_null_industries = pd.read_sql_query(query_5_4_1, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [212]:
# текст запроса
query_5_5 = f'''select
                e.name,
                count(e.id)
                from employers e
                join employers_industries ei on ei.employer_id = e.id
                group by 1
                having count(e.id) = 4
                order by 1
'''

In [213]:
# результат запроса
employers_with_four_industries = pd.read_sql_query(query_5_5, connection)
print(employers_with_four_industries)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\3801801752.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_with_four_industries = pd.read_sql_query(query_5_5, connection)


                 name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [214]:
# текст запроса
query_5_6 = f'''select
                count(ei.employer_id)
                from employers_industries ei
                join industries i on i.id = ei.industry_id
                where i.name  like 'Разработка программного обеспечения'
'''

In [215]:
# результат запроса
employers_of_developers= pd.read_sql_query(query_5_6, connection)
print(employers_of_developers)

   count
0   3553


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\2916004208.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_of_developers= pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [216]:
# код для получения списка городов-милионников

In [217]:
# текст запроса

In [218]:
# результат запроса

***

##### Выводы
1. Первые 4 места по количеству вакансий занимают IT, фин-тех. гиганты такие как Яндекс, Тинькофф, Ростелеком, Сбер. Это говорит о стремительном развитии сферы высоких, цифровых технологий, которые проникают во все сферы жизнедеятельности и требуют специалистов.
2. В базе присутствуют регионы с работодателями, но без вакансий. Вероятно, это некорректно завершенные регистрации работодателей.
3. Наибольшее присутствие по регионам у таких  компаний как Яндекс, который видимо нанимает работников на удаленные вакансии, Ростелекома, имеющего представительства по всей России. Компании Сецремонт, веротятно набирающей работников вахтовым методом.
4. Более 8 тыс работодателей не указали сферу деятельности. Это компании из разных отраслей судя по названиям IT, бюджетные и прочие. Связано это, видимо, с вводом неполных данных при регистрации.
5. Доля вакансий для разработчиков составляет почти 10% от общего количества вакансий. Это подтверждает и дополняет предыдущие выводы о значительном количестве вакансий и работодателей связанных с IT отраслью.




# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [219]:
# текст запроса
query_6_1 = f'''select
                count(id)
                from vacancies 
                where lower(name) like '%data%' 
                or lower(name) like '%данн%'
'''

In [220]:
# результат запроса
count_of_data_vacancies = pd.read_sql_query(query_6_1, connection)
print(count_of_data_vacancies)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\701109852.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_data_vacancies = pd.read_sql_query(query_6_1, connection)


   count
0   1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [221]:
# текст запроса
query_6_2 = f'''select
                count(id)
                from vacancies
                where (lower(name) like '%data scientist%' or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%')
                or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
                and ((lower(name) like '%junior%') or (experience = 'Нет опыта') or (employment = 'Стажировка'))
'''

In [222]:
# результат запроса
count_of_DSJ_vacancies = pd.read_sql_query(query_6_2, connection)
print(count_of_DSJ_vacancies)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\707981503.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_DSJ_vacancies = pd.read_sql_query(query_6_2, connection)


   count
0     51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [223]:
# текст запроса
query_6_3 = f'''select
                count(id)
                from vacancies
                where (lower(name) like '%data scientist%' or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%')
                or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
                and ((lower(key_skills) like '%sql%') or (lower(key_skills) like '%postgres%'))
'''

In [224]:
# результат запроса
count_of_DS_vacancies_with_sql = pd.read_sql_query(query_6_3, connection)
print(count_of_DS_vacancies_with_sql)

C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1784486223.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_DS_vacancies_with_sql = pd.read_sql_query(query_6_3, connection)


   count
0    201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [225]:
# текст запроса
query_6_4 = f'''select
                count(id)
                from vacancies
                where (lower(name) like '%data scientist%' or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%')
                or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
                and (lower(key_skills) like '%python%')
'''

In [226]:
# результат запроса
count_of_DS_vacancies_with_python = pd.read_sql_query(query_6_4, connection)
print(count_of_DS_vacancies_with_python)

   count
0    351


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1904482588.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_DS_vacancies_with_python = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [227]:
# текст запроса
query_6_5 = f'''select
                round(avg(length(key_skills)-length(replace(key_skills, CHR(9),''))), 2)+1
                from vacancies
                where (lower(name) like '%data scientist%' or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%')
                or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
                and key_skills is not null
'''

In [228]:
# результат запроса
count_of_keyskills_DS = pd.read_sql_query(query_6_5, connection)
print(count_of_keyskills_DS)


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\4117159953.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_keyskills_DS = pd.read_sql_query(query_6_5, connection)


   ?column?
0      6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [229]:
# текст запроса
query_6_6 = f'''select
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) as avg_salary,
                experience
                from vacancies
                where (lower(name) like '%data scientist%' or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%' or (name like '%ML%' and name not like '%HTML%')
                or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%')
                and (salary_from is not null or salary_to is not null)
                group by experience
'''

In [230]:
# результат запроса
avg_salary_of_DS = pd.read_sql_query(query_6_6, connection)
print(avg_salary_of_DS)

   avg_salary          experience
0     74643.0           Нет опыта
1    139675.0  От 1 года до 3 лет
2    243115.0       От 3 до 6 лет


C:\Users\pg-nv\AppData\Local\Temp\ipykernel_820\1999635611.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary_of_DS = pd.read_sql_query(query_6_6, connection)


***

##### Выводы
1. Количество вакансий связанных с DS 1771 шт. это ак же значительный объм, составляющий почти 5% от количества вакансий.
2. Для специалистов уровня Junior количество вакансий невелико всего 51
3. Ключевые навыки специалистов DS это Python, SQL. В среднем от датасайентиста требуют знаний по более чем 6 ключевым навыкам.
4. Для вакансий в сфере DS работодатели предлагают зарплаты около 74 тыс.руб. для новичков (junior), около 140 тысяч для специалистов с опытом от 1 года до 3-х лет. И 243 тыс. руб. для датасайентиста с опытом от 3-х до 6 лет.


# Общий вывод по проекту

Для трудового рынка России и СНГ, на основе базы данных работодателей HH.RU можно сделать следующие выводы:
- Основные работодатели на HH.RU это российские финтех, IT гиганты Яндекс,Сбер,Тинькофф, Ростелеком
- Доля IT специалистов в вакансиях велика. Например, разработчики и специалисты DS это более 5300 выкансий из 49 тыс. Можно дополнить анализ запросом, который выведет всех специалистов IT, таких как тестировщики, UX/UI дизайнеры, аналитики и др.
- Спрос на удаленных работников, так же высок. В совокупности с растущим спросом на IT специалистов, это подверждает рост этой отрасли.
- При анализе вакансий специалистов DS можно сделать вывод, что работодателям требуются специалисты с имеющимся опытом от 1 года до 3-х лет. Это характерно и для всех вакансий в базе. Выйти на рынок для навичков, очень сложно. Необходимо обладать широким спектром навыков.
- Для специалистов IT, для датасайентистов в частности, характерен высокий уровень зарплат. Новичкам, готовы платить среднюю зарплату на рынке вакансий, а для специалистов с опытом, уровень предлягаемых зарплат в 2 и более раз выше, чем средний уровень зарплат по стране.

In [232]:
connection.close()